File: 95. Unique Binary Search Trees II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def generateTrees(self, n: int) -> List[TreeNode]:
        if n == 0:
            return []
        return self.helper(1, n)
        
    def helper(self, start, end):
        if start > end:
            return [None]
        res = []
        for i in range(start, end+1):
            left_part = self.helper(start, i-1)
            right_part = self.helper(i + 1, end)
            
            for l in left_part:
                for r in right_part:
                    root = TreeNode(i)
                    root.left = l
                    root.right = r
                    res.append(root)
        return res

File: 95. Unique Binary Search Trees II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    /**
     * Recursion to pick a root and genreate all possible trees,
     * then append it to the list
     * Time O(4^n/(n^0.5)) since generate a tree with n nodes are Cn = Sum(Ci * Cn-i) and we are generating n times
     * Space also O(4^n/(n^0.5))
     * @param n
     * @return
     */
    public List<TreeNode> generateTrees(int n) {
        if (n == 0) {
            return new LinkedList<TreeNode>();
        }
        return generate(1,n);
    }
    
    private List<TreeNode> generate(int st, int ed) {
        List<TreeNode> trees = new LinkedList<TreeNode>();
        if (st > ed) {
            trees.add(null);
            return trees;
        }
        
        for (int i=st;i<=ed;i++) {
            List<TreeNode> leftTrees = generate(st, i-1);
            List<TreeNode> rightTrees = generate(i+1, ed);
            for (TreeNode l : leftTrees) {
                for (TreeNode r : rightTrees) {
                    TreeNode treeRoot = new TreeNode(i);
                    treeRoot.left = l;
                    treeRoot.right = r;
                    trees.add(treeRoot);
                }
            }
        }
        return trees;
    }
}